Compass setting not read/save properly

DarkAngelFR opened the pull request:
#Fixes:
- Compass settings not saved properly

#Changes:
- N/A

#Change details:
- N/A

#Other informations:
- N/A

advancedrei commented:
Oops, sorry about that. Good catch :).

DarkAngelFR commented:
@advancedrei with previous code, timer was only started on resuming the gamepage. also the parameter autorotate was used to trigger or not the event making maps very strange : you need rotate + compass activated to have compass but rotate based on headings was giving map movement very strange (mix of heading + compass ^^)

advancedrei commented:
So, part of the reason I shut it off was the code that was added 4 days ago started draining my battery faster and making my phone run hotter. With the changes you made, this PR doesn't fix THAT problem.

@st-apps my personal opinion is that you should not accept this PR yet.

Can someone go in and look at what the compass code (if there was any) looked like > 4 days ago? I don't have time to look today, but a) the map behavior once the compass was added made it unusable for me, and b) the official app does not autorotate the map based on the compass, it only does so during forward motion. Given all that, I think the Settings feature should let you shut the compass all the way off, unless it is **absolutely** required to drive the map.

DarkAngelFR commented:
ok I will update my PR with a stop of the compass timer when setting is stopped (should be quick by registering for propchange on the setting)
will do this tonight as I'm entering a meeting now and won't have time for this :(

@ST-Apps do you prefer me to close this one and do a fresh new one (using merge) ?

DarkAngelFR commented:
I updated the PR
@advancedrei could you take a little time to test battery use ? on my phone it looks ok....
@ST-Apps I let you have a look at the change, I updated a little the settings instance ;)

advancedrei commented on PokemonGo-UWP/Utils/GameClient.cs:
```diff
@@ -490,25 +491,35 @@ public static AccessToken LoadAccessToken()
         {
-            if (SettingsService.Instance.IsCompassEnabled)
+            #region Compass management
+            _compass = Compass.GetDefault();
```
Here's my question: Is turning the compass on and querying it for data *absolutely necessary* for the map to operate better? To me, it worked just fine without it. If the user doesn't want the compass turned on, it shouldn't even create an instance of the object and start that power consumption, **_unless_** that data is needed to provide the correct initial map view.

ST-Apps replied:
I think that this has been added for cases in which the GPS doesn't return the current `Heading`, so that users can still have the map rotate as they move.

advancedrei replied:
Ok, so then in this case I think the compass-related functions should stay wrapped in the settings check per my version on the left. That way, nothing is even accessed if the user does not wish it. Saves battery and memory allocations.

DarkAngelFR replied:
Idea was to have 2 different rotation option, one using GPS the other one using compass, you are right using both is a non-sense.
GPS only is good when walking, compas is better when standing or walking slowly (to my opinion)

advancedrei replied:
Well, it's not that having two is nonsense. It's that having both on at the same time causes the map to be **totally useless**. It flips and spins and you can't do anything. I was trying to KISS by just letting you shut it off. If you want it to be smarter, then it should only be querying for compass data if you've identified after several GPS coordinate returns that you're getting velocity information but not a heading.

BUT IMHO these settings serve different purposes, and how they are named is fine. Let's clean up the logic and worry about the naming later.

advancedrei commented on PokemonGo-UWP/Utils/GameClient.cs:
```diff
@@ -490,25 +491,35 @@ public static AccessToken LoadAccessToken()
                     _compassTimer.Start();
-                }
+
+                SettingsService.Instance.PropertyChanged += (object sender, PropertyChangedEventArgs e) => 
```
This is definitely needed regardless, but I think it would be much better of there were simple "CompassOn" and "CompassOff" functions that manage the entire Compass instance, not just the timer. That way, the variables can be garbage collected if they aren't being used.

DarkAngelFR replied:
true, not sure if it's the good place to put this (GameClient is more focused on connection between game and niantics calls). Maybe having an instance dedicated to this should be better (could have gps, compass, gyroscope(?), other "device orientation and location" stuf)
any suggestion? (in waiting for a clean refactoring I will try to keep things as much as possible in this init region)

advancedrei replied:
Just remember, the official app only uses the compass for the AR stuff. So it appears you're trying to engineer for unnecessary things. My advice would be to focus on fixing the problem at hand and don't worry about creating separate helpers or moving the code or what the UI might do with it later. The original problem was that the ViewModel wasn't updating the SettingService. The subsequent changes confused my original intent of allowing you to shut a feature that often isn't needed off so it doesn't allocate memory or battery.

advancedrei commented on PokemonGo-UWP/Views/GameMapPage.xaml.cs:
```diff
@@ -221,7 +221,8 @@ private void SubscribeToCaptureEvents()
         private void HeadingUpdated(object sender, Windows.Devices.Sensors.CompassReading e)
         {
-            GameMapControl.Heading = e.HeadingTrueNorth ?? e.HeadingMagneticNorth;
+            if (SettingsService.Instance.IsCompassEnabled)
```
If this is inside a compass event handler, is this check necessary? It shouldn't even be firing if the compass setting is off. Otherwise the compass is still using battery power.

DarkAngelFR replied:
when testing, it turns out that between the uncheck and the property change event, the timer ticks (one or 2 times max), this "if" is just here to avoid it but not really necessary

advancedrei replied:
Might consider adjusting the timer to be less sensitive.

advancedrei commented on PokemonGo-UWP/Utils/Settings/SettingsService.cs:
```diff
@@ -2,22 +2,20 @@
 namespace PokemonGo_UWP.Utils
 {
-    public class SettingsService
+    public class SettingsService: BindableBase
```
This is a good change. We should probably make the whole SettingsService bindable and clean up this file. We could delete a lot of code in the SettingsPageViewModel that way.

DarkAngelFR replied:
I will refactor/clean the settings stuff for next PR ;) BindableBase is usefull mixed with the helper should be nice (BindableSettingsHelper lol)

advancedrei replied:
Well, I think you should consider holding off on that just yet. I think we should get 1.1 shipped, and then after that look at how are we going to use something like Dependency Injection to clean up some of these types of issues. Having a single bindable Settings instance injected into every ViewModel that needs it would eliminate some unnecessary bridge properties and make things a lot cleaner.

advancedrei commented on PokemonGo-UWP/Utils/GameClient.cs:
```diff
@@ -490,25 +491,35 @@ public static AccessToken LoadAccessToken()
+                };
+                _compassTimer.Tick += (s, e) =>
+                {
+                    if (SettingsService.Instance.IsCompassEnabled) //avoid reading compass value if compass is off, should'nt be needed as we stop the compass tick on prop change but who knows!
```
On Line 504 and 507, you're checking to see if the compass is on, but you removed the AutoRotateMap check, which is also part of this process. No sense reading the compass if they're not auto-rotating the map.

DarkAngelFR replied:
idea was to send events to the page so the UI takes care of it (use it or not, register or not, ...)
goal was to avoid having an event not being called because of some UI stuff (if someone one day wants to use this event somewhere else, he would have to be carefull that the event is not always trigger which is seems strange to me ^^ )

DarkAngelFR commented:
hey both ;)
I read comments, you're right some stuff are not perfect...
I will clean-up all this tomorrow morning and PR here.
Quickly :
- I will change for creating compass stuff only if compass is activated by user on settings
- I will rename a little the 2 settings : AutoRotate for rotation, if checked then 2 options = rotate based on GPS move (headings) or rotate based on compass (using compass reader), maybe later we can add also a rotate based on gyroscope (to have map going up and down ^^)
- I will clean the settings stuff to be clear for notification stuff

Hope it will be more clean this way ;)
